nexus-repository-r/src/main/java/org/sonatype/nexus/repository/r/internal/RPackagesBuilderFacetImpl.java

```diff
@@ -0,0 +1,198 @@
+    if (shouldInvalidate(deleted)) {
+      invalidateMetadata(getBasePath(deleted.getAsset().name()));
+    }
+  }
+
+  /**
+   * Handles {@code AssetCreatedEvent} events concurrently, requesting a metadata invalidation and rebuild if warranted
+   * by the event contents.
+   *
+   * @param created The creation event to handle.
+   */
+  @Subscribe
+  @Guarded(by = STARTED)
+  @AllowConcurrentEvents
+  public void on(final AssetCreatedEvent created) {
+    if (shouldInvalidate(created)) {
+      invalidateMetadata(getBasePath(created.getAsset().name()));
+    }
+  }
+
+  /**
+   * Handles {@code AssetUpdatedEvent} events concurrently, requesting a metadata invalidation and rebuild if warranted
+   * by the event contents.
+   *
+   * @param updated The update event to handle.
+   */
+  @Subscribe
+  @Guarded(by = STARTED)
+  @AllowConcurrentEvents
+  public void on(final AssetUpdatedEvent updated) {
+    if (shouldInvalidate(updated)) {
+      invalidateMetadata(getBasePath(updated.getAsset().name()));
+    }
+  }
+
+  /**
+   * Handles metadata invalidation for this repository given the specified base path. Internally this is handled by
+   * posting a {@code RMetadataInvalidationEvent} to the event manager which then triggers the actual metadata
+   * regeneration. Note that the regeneration may not be instantaneous (and that the waiting period is configurable).
+   *
+   * @param basePath The base path of the PACKAGES.gz file to invalidate.
+   */
+  @Override
+  public void invalidateMetadata(final String basePath) {
+    if (!waiting) {
+      eventManager.post(new RMetadataInvalidationEvent(getRepository().getName(), basePath));
+    }
+  }
+
+  /**
+   * Listen for invalidation of the metadata, wait a configured time and then rebuild. The waiting allows throwing away
+   * of subsequent events to reduce the number of rebuilds if multiple archives are being uploaded. This method must NOT
+   * be allowed to process concurrent events as race conditions may result when rebuilding the data.
+   */
+  @Subscribe
+  public void on(final RMetadataInvalidationEvent event) {
+    if (getRepository().getName().equals(event.getRepositoryName())) {
+      try {
+        waiting = true;
+        Thread.sleep(interval);
+      }
+      catch (InterruptedException e) {
+        log.warn("R invalidation thread interrupted on repository {}, proceeding with invalidation",
+            getRepository().getName());
+        Thread.currentThread().interrupt();
+      }
+      waiting = false;
+      log.info("Rebuilding R PACKAGES.gz metadata for repository {}", getRepository().getName());
+      UnitOfWork.begin(getRepository().facet(StorageFacet.class).txSupplier());
+      try {
+        RHostedFacet hostedFacet = getRepository().facet(RHostedFacet.class);
+        hostedFacet.buildAndPutPackagesGz(event.getBasePath());
+      }
+      catch (IOException e) {
+        throw new UncheckedIOException(e);
+      }
+      finally {
+        UnitOfWork.end();
```

Comment:
UnitOfWork is AutoClosable - why don't you use it in 'try with resources'?

Reply:
done in fe40919080a00ea9162d0af9be02c414af338e27

Reply:
I cannot add `UnitOfWork.begin` into try with resources